build-system/babel-config/pre-closure-config.js

```diff
@@ -49,6 +49,7 @@ function getPreClosureConfig() {
   ];
   const replacePlugin = getReplacePlugin();
   const preClosurePlugins = [
+    argv.coverage ? './build-system/babel-plugins/babel-plugin-istanbul' : null,
```

Comment:
Our compilation pipeline already has the capability to instrument code with `istanbul` during unminified builds (see https://github.com/ampproject/amphtml/commit/905aa951bef81f5c0f38891e585d88d1d1aa5ed8). I wonder if you can simply re-use that logic for minified code without having to do everything this PR does.

To be more specific, I wonder if you can insert a copy of this line...

https://github.com/ampproject/amphtml/blob/7367cb48eddc0c0809563f1d595b0817985b2adb/build-system/tasks/helpers.js#L395

... somewhere in this set of lines (maybe just before line 354?)...

https://github.com/ampproject/amphtml/blob/7367cb48eddc0c0809563f1d595b0817985b2adb/build-system/compile/compile.js#L352-L381

Reply:
After adding that flag, do you know if you/anyone else ever added anything that uses the resulting coverage data? As far as I can tell, it instruments post-compilation/bundle/transformation code, meaning the coverage represents lines in the binaries not the original source files. The coverage itself is stored in a variable of the form `$$cov_{timestamp}$$`, which would have to be extracted from the page somehow (I know jest and karma have some built-in support for this, but it doesn't typically require doing the instrumentation yourself either). The resulting coverage data does not match anything reported to Codecov, and so far I haven't found anything that attempts to access the coverage variables, but I could be missing something.

Sample of coverage data from `gulp build --coverage --fortesting && gulp serve --coverage`:
![image](https://user-images.githubusercontent.com/6694512/91606569-4ac7ac80-e940-11ea-9a0a-605ea31bdc3c.png)

Based on the digging I've done, it seems to me like the gulp plugin won't work for our use-case because A) if it runs before compilation, it will fail trying to instrument unsupported ES6 features and B) if it runs after compilation, the instrumentation results in less useful coverage data of lines at the binary level (possibly after tree-shaking, skewing the results even further). The babel plugin appears to interop with the other transpilation steps without issue.

Sample of coverage data from `gulp dist --coverage && gulp serve --coverage --compiled`:
![image](https://user-images.githubusercontent.com/6694512/91609083-ba3f9b00-e944-11ea-80dd-dee750d499c8.png)
![image](https://user-images.githubusercontent.com/6694512/91609098-c3c90300-e944-11ea-9d3c-5f2041057698.png)

Reply:
> After adding that flag, do you know if you/anyone else ever added anything that uses the resulting coverage data?

High level answer: no. I only managed to add instrumentation for unminified runtime code, and never actually used the results in tests (manual or automated). For minified code, the last time I checked, adding instrumentation caused a significant performance hit to `gulp dist`. I'm hoping things are different now with recent changes to our compilation toolchain (native closure compiler) and the use of babel plugins (your changes in this PR).

> Based on the digging I've done, it seems to me like the gulp plugin won't work for our use-case because A) if it runs before compilation, it will fail trying to instrument unsupported ES6 features and B) if it runs after compilation, the instrumentation results in less useful coverage data of lines at the binary level (possibly after tree-shaking, skewing the results even further). The babel plugin appears to interop with the other transpilation steps without issue.

In that case, I'm fully supportive of ripping out the use of the `gulp` plugin for `istanbul` and replacing it with the babel plugin you're introducing. (I'd recommend directly installing `babel-plugin-istanbul` to `node_modules`).

Reply:
It's my understanding that instrumentation is effectively guaranteed to increase `gulp dist` time significantly, as it nearly doubles the number of instructions being minified (since the instruction counter code itself is inserted almost everywhere). I suspect this is unavoidable, though the babel plugin itself _might_ offer _slightly_ better performance if it's optimized in anticipation of minification. All of our karma code is instrumented automatically IIRC, and that probably accounts for a significant chunk of the time spent building test code too

Reply:
WIP: Migrating existing `gulp build --coverage` to babel plugin instead of gulp plugin

Reply:
@rsimha I've migrated `gulp build` to use the babel plugin instead of the gulp one, and ran some tests. For both `gulp build` and `gulp dist`, adding `--coverage` roughly quadruples the time required. For the reasons mentioned above, it's not surprising this happens, and I suspect it's mostly unavoidable given how code instrumentation operates.

One piece of future investigatory work worth exploring might be attempting to disable minification/obfuscation of the actual coverage variables inserted, as maybe this would save some time for babel/closure. It's hard to know what a reasonable expectation is given there's no solid baseline for compiling with instrumentation. Down the line, we can use tools like `yarn 0x` to generate flame charts of the build/dist process, and try to identify which steps incur the greatest differential cost and selectively try to optimize them.

| Build Times | Standard | With `--coverage` |
| --- | --- | --- |
| `gulp build --fortesting` | 4.53m | 16m |
| `gulp dist --fortesting` | 4.22m | 20m |

There is no significant difference in the runtime of `gulp build --fortesting` between the gulp plugin implementation and the babel plugin version, but the output with the babel version is preferable since it references actual source file rather than coverage of the compiled binaries.

Reply:
Thanks for the analysis. Agree, it would be good to further improve instrumented compilation time. Meanwhile, this is still a substantial improvement over previous times (standard: ~15 min, instrumented: ~60 min). Since this PR as written isn't going to affect CI round trip time, I'm going to resolve this thread and approve the PR.

Reply:
Agreed, I'm investigating source map support (there are literally dozens of istanbul-related packages, with varying levels of support). My plan is to get this is to unblock work on coverage collection for manual/e2e testing, and in parallel try to move the instrumentation to the end of the transpilation chain

Reply:
Update: In doing some digging, I found that **both** the babel and gulp plugins support sourcemaps, but only when inlined. I think the only reason they weren't being used is because that call to `istanbul()` comes before the sourcemaps are loaded from the generated `.map` files (which the plugins themselves don't support). Doing some experimenting, but if it pans out the way I hope, we may not need pre-instrumentation at all :astonished:

Reply:
Epilogue: running `nyc instrument` itself on the output of `build` or `dist` takes up the difference in time, ie. it's not wasted work due to minifying coverage code, it's just genuinely expensive to instrument code no matter what.